File: PolyScript/frameworks/test/test-compiler.py
#!/usr/bin/env python3

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))

from polyscript_click import polyscript_tool

@polyscript_tool()
class TestCompilerTool:
    """Test compiler tool for validating CRUD × Modes framework"""
    
    def create(self, resource, options, context):
        """Create operation - compile source files"""
        context.log(f"Creating compilation target: {resource}")
        
        output_file = options.get("output") or resource.replace(".py", ".pyc")
        
        return {
            "operation": "create",
            "compiled": resource,
            "output": output_file,
            "optimized": options.get("optimize", False),
            "timestamp": context.output_data["data"].get("timestamp", "now"),
            "mode": context.mode
        }
    
    def read(self, resource, options, context):
        """Read operation - show compilation status"""
        context.log("Checking compilation status...")
        
        files = [resource] if resource else ["main.py", "utils.py", "config.py"]
        
        return {
            "operation": "read",
            "source_files": files,
            "compiled_files": [f.replace(".py", ".pyc") for f in files[:-1]],
            "missing": [files[-1].replace(".py", ".pyc")],
            "last_build": "2025-01-04T12:00:00Z",
            "mode": context.mode
        }
    
    def update(self, resource, options, context):
        """Update operation - recompile changed files"""
        if not resource:
            raise ValueError("Resource is required for update operation")
            
        context.log(f"Recompiling {resource}...")
        
        return {
            "operation": "update",
            "recompiled": resource,
            "reason": "source file changed",
            "timestamp": "2025-01-04T12:00:00Z",
            "incremental": options.get("incremental", False),
            "mode": context.mode
        }
    
    def delete(self, resource, options, context):
        """Delete operation - clean build artifacts"""
        context.log(f"Cleaning build artifacts{' for ' + resource if resource else ''}...")
        
        targets = [f"{resource}.pyc"] if resource else ["*.pyc", "__pycache__/", "dist/"]
        
        return {
            "operation": "delete", 
            "cleaned": targets,
            "freed_space": "15.3 MB",
            "timestamp": "2025-01-04T12:00:00Z",
            "mode": context.mode
        }
    
    def add_arguments(self, cmd):
        """Add tool-specific arguments"""
        # For Click, we need to add options using decorators or params
        # This is handled by the framework

if __name__ == "__main__":
    TestCompilerTool.run()

File: PolyScript/frameworks/test/TestCompiler.hs
{-|
Test Haskell Compiler Tool for PolyScript Framework
CRUD × Modes Architecture: Zero-boilerplate CLI development

Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
-}

{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

module Main where

import Control.Monad.IO.Class (liftIO)
import Data.Aeson (object, (.=), Value(..), ToJSON)
import qualified Data.Aeson as Aeson
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time (getCurrentTime, formatTime, defaultTimeLocale)
import GHC.Generics (Generic)
import PolyScript.Framework

-- Test Compiler Tool Implementation
data TestCompilerTool = TestCompilerTool

instance PolyScriptTool TestCompilerTool where
  toolDescription _ = "Test Haskell compiler tool demonstrating CRUD × Modes"
  
  toolCreate _ resource options context = do
    logMessage context "info" $ "Compiling " <> T.pack (show resource) <> "..."
    
    let resourceStr = maybe "main.hs" T.unpack resource
    let outputFile = maybe (T.pack $ resourceStr ++ ".out") T.pack (lookup "output" options >>= parseString)
    let optimize = maybe False parseBool (lookup "optimize" options)
    
    currentTime <- getCurrentTime
    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%z" currentTime
    
    return $ Right $ object
      [ "compiled" .= resourceStr
      , "output" .= outputFile
      , "optimized" .= optimize
      , "timestamp" .= timestamp
      ]
  
  toolRead _ resource options context = do
    logMessage context "info" "Checking compilation status..."
    
    let sourceFiles = case resource of
          Just r -> [T.unpack r]
          Nothing -> ["Main.hs", "Utils.hs", "Config.hs"]
    
    currentTime <- getCurrentTime
    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%z" currentTime
    
    return $ Right $ object
      [ "source_files" .= sourceFiles
      , "compiled_files" .= (["Main.hi", "Utils.hi"] :: [String])
      , "missing" .= (["Config.hi"] :: [String])
      , "last_build" .= timestamp
      ]
  
  toolUpdate _ resource options context = do
    logMessage context "info" $ "Recompiling " <> T.pack (show resource) <> "..."
    
    let resourceStr = maybe "main.hs" T.unpack resource
    let incremental = maybe False parseBool (lookup "incremental" options)
    
    currentTime <- getCurrentTime
    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%z" currentTime
    
    return $ Right $ object
      [ "recompiled" .= resourceStr
      , "reason" .= ("source file changed" :: Text)
      , "incremental" .= incremental
      , "timestamp" .= timestamp
      ]
  
  toolDelete _ resource options context = do
    logMessage context "info" $ "Cleaning " <> maybe "build artifacts" (T.pack . show) resource <> "..."
    
    currentTime <- getCurrentTime
    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%z" currentTime
    
    return $ Right $ object
      [ "cleaned" .= (["*.hi", "*.o", "dist/"] :: [String])
      , "freed_space" .= ("8.9 MB" :: Text)
      , "timestamp" .= timestamp
      ]

-- Helper functions
parseString :: Aeson.Value -> Maybe String
parseString (String t) = Just $ T.unpack t
parseString _ = Nothing

parseBool :: Aeson.Value -> Bool
parseBool (Bool b) = b
parseBool _ = False

main :: IO ()
main = runTool TestCompilerTool